src/libraries/Microsoft.Extensions.Diagnostics/src/Metrics/DefaultMeterFactory.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace Microsoft.Extensions.Diagnostics.Metrics
+{
+    internal sealed class DefaultMeterFactory : IMeterFactory
+    {
+        private readonly Dictionary<string, List<Meter>> _cachedMeters = new();
+        private bool _disposed;
+
+        public DefaultMeterFactory() { }
+
+        public Meter Create(MeterOptions options)
+        {
+            if (options is null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            Debug.Assert(options.Name is not null);
+
+            lock (_cachedMeters)
+            {
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(nameof(DefaultMeterFactory));
+                }
+
+                if (_cachedMeters.TryGetValue(options.Name, out List<Meter>? meterList))
+                {
+                    foreach (Meter meter in meterList)
+                    {
+                        if (meter.Version == options.Version && DiagnosticsHelper.CompareTags(meter.Tags, options.Tags))
```

Comment:
I think this is sensitive to tag order. Has that been considered?

Reply:
Isn't the tag order should be sensitive? 

@dpk83 do you have input on that?

Reply:
I'll open a new issue to track discussing this as we shouldn't block this PR on that.

Reply:
What I'm getting at is that if you specify meter options with the same name and two tags in different `MeterFactory.Create` calls:
* Create(Name = "Test", Tags = "A=1, B=2")
* Create(Name = "Test", Tags = "B=2, A=1")
 
The result is two different meters. I don't know if it's important or not. Do exporters, e.g. Prometheus, care about tag order?

Reply:
> The result is two different meters. Do exporters, e.g. Prometheus, care about tag order?

CC @reyang if knows about that.

Reply:
I think the order shouldn't matter (e.g. "A=1, B=2" should be treated the same way as "B=2, A=1"). What's more, I suggest the following flow:
1. order the attributes by names
2. if there is any duplicates (e.g. the same key value specified more than once, the same name is used with different combination of tags or other values), have a well defined behavior (e.g. either keep silent and let the SDK handle it, or raise exception and let the creation fail).

Reply:
Looks out metrics aggregator is sorting them which suggest the order doesn't matter. I mean the tags are order insensitive. 

https://github.com/dotnet/runtime/blob/8163d1643afe2c029c1783b5f92548e6dc6828ff/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs#L335

CC @noahfalk

Reply:
Everywhere tags are discussed in OTel spec and in .NET docs we should be saying it is order insensitive. I'd also recommend we make this caching use order insensitive matching.

Reply:
I'll address this in a separate PR and we already having https://github.com/dotnet/runtime/issues/86614 for tracking it.